compiler/passes/src/write_transforming/expression.rs

```diff
@@ -0,0 +1,261 @@
+            // Build the struct expression from the members.
+            let id = self.state.node_builder.next_id();
+            self.state.type_table.insert(id, ty.clone());
+            let Type::Composite(comp_type) = ty else {
+                panic!("The type of a struct init should be a composite.");
+            };
+            let expr = StructExpression {
+                members: struct_members
+                    // This clone is unfortunate, but both `struct_members` and the closure below borrow self.
+                    .clone()
+                    .iter()
+                    .map(|(field_name, ident)| {
+                        let (expr, statements2) = self.reconstruct_identifier(*ident);
+                        statements.extend(statements2);
+                        StructVariableInitializer {
+                            identifier: Identifier::new(*field_name, self.state.node_builder.next_id()),
+                            expression: Some(expr),
+                            span: Default::default(),
+                            id: self.state.node_builder.next_id(),
+                        }
+                    })
+                    .collect(),
+                name: comp_type.id,
+                span: Default::default(),
+                id,
+            };
+            let statement = AssignStatement {
+                place: input.into(),
+                value: expr.into(),
+                span: Default::default(),
+                id: self.state.node_builder.next_id(),
+            };
+            statements.push(statement.into());
+            (input.into(), statements)
+        } else {
+            // This is not a struct or array whose members are written to, so there's nothing to do.
+            (input.into(), Default::default())
+        }
+    }
+
+    fn reconstruct_array_access(&mut self, input: ArrayAccess) -> (Expression, Self::AdditionalOutput) {
+        let Expression::Identifier(array_name) = input.array else {
+            panic!("SSA ensures that this is an Identifier.");
+        };
+        if let Some(member) = self.get_array_member(array_name.name, &input.index) {
+            self.reconstruct_identifier(member)
+        } else {
+            (input.into(), Default::default())
+        }
+    }
+
+    fn reconstruct_member_access(&mut self, input: MemberAccess) -> (Expression, Self::AdditionalOutput) {
+        let Expression::Identifier(array_name) = input.inner else {
+            panic!("SSA ensures that this is an Identifier.");
+        };
+        if let Some(member) = self.get_struct_member(array_name.name, input.name.name) {
+            self.reconstruct_identifier(member)
+        } else {
+            (input.into(), Default::default())
+        }
+    }
+
+    // The rest of the methods below don't do anything but traverse - we only modify their default implementations
+    // to combine the `Vec<Statement>` outputs.
```

Comment:
Not needed in this PR, but we could prolly add bounds on `Self::AdditionalOutput` so that the default impl has some notion of combining output.